### PR TITLE
feat(bamlparser): add shared BAML parser package side-by-side (#265 PR 1)

### PR DESCRIPTION
**Pull request by invakid404**
## Summary

First of four PRs unifying baml-rest's two in-tree BAML parsers (#265). Adds a shared `bamlutils/bamlparser` package side-by-side with the existing parsers, plus a dual-parser parity test harness. **No production code changes** — `cmd/introspect/main.go` and `bamlutils/versions.go` still use their existing parsers; the new package is consumed only by the parity tests.

Refs #265. Umbrella stays open until PR 4.

## What's included

- New `bamlutils/bamlparser/` package: participle-lexer-driven AST parser + AST types + value helpers.
- AST captures literal-vs-`env.X` provenance at parse time as a first-class `Value` node — preserves the `IsSet()` declared-presence semantics from #263.
- Permissive `Other`/`TypeBlock` items tolerate upstream constructs (class, enum, type alias, template_string, test) that introspect currently ignores — the new parser must not error on them.
- Unit tests (`bamlutils/bamlparser/bamlparser_test.go`) pinning the parser's standalone contract: AST round-trips, lexer edge cases (raw strings, `env.X` vs quoted `"env.X"`, comments inside strings), `Other`-block tolerance for nested generics, compact one-line blocks, top-level garbage robustness.
- Dual-parser parity test (`cmd/introspect/baml_parser_parity_test.go`): every fixture from `cmd/introspect/baml_parser_test.go` plus hand-built edge cases enumerated during scoping (65 cases total). Asserts identical normalized `bamlConfig` output between old and new parsers. `knownParityDivergences` is empty for this PR — any divergence is a correctness issue, not a known difference.

## Sequence

This is PR 1/4:
- PR 1 (this) — introduce shared parser side-by-side
- PR 2 — switch `cmd/introspect` to use the new parser (parity-gated)
- PR 3 — delete old parser helpers + retarget helper-specific tests
- PR 4 — migrate `bamlutils/versions.go`

## Parity-with-upstream (tracked separately)

Behavior gaps between baml-rest's parser and upstream BAML (plain `client Name { }` blocks, retry precedence quirk, case-sensitivity, strategy list whitespace splitting) are intentionally NOT addressed in this refactor sequence. They are tracked at #268 as deliberate behavior changes to land as individual follow-up PRs once #265 completes. This refactor preserves current behavior exactly.

## Notes

- `bamlutils/embed.go` was regenerated by `go run ./cmd/embed` to include the new package's source files (which the runtime embeds for diagnostic display).
- `go.work.sum` picked up one indirect dependency hash from `./scripts/sync.sh` — no version pin changes.

## Verification

- `go build ./...` / `go vet ./...` clean.
- `go test ./bamlutils/bamlparser -count=1 -v` — green.
- `go test ./cmd/introspect -count=1 -v` — existing tests green + 65-case parity table green.
- `go test ./...` green.
- `verify-framework-adapter`: 6/6 (post-PR 4b shape).
- `verify-adapter-pins`: 4/4.
- `./scripts/sync.sh` clean for the gated paths (`git diff --exit-code -- adapters/adapter_v0_*/go.{mod,sum} adapters/common/go.{mod,sum}` clean).

Refs #265.

## Summary by CodeRabbit

* **Refactor**
  * Introduced an AST-driven BAML parser for more robust, ordered handling of top-level blocks and nested bodies.
* **New Features**
  * Improved value handling (quoted/raw strings, env refs, numbers, identifiers, lists) with helpers for string/list extraction.
  * New parsing entry points for loading BAML from files/strings/readers.
* **Tests**
  * Extensive unit and parity tests covering many syntaxes and ensuring compatibility with the legacy parser.
* **Chores**
  * Updated embedded sources list for tooling consistency.

[Review Change Stack](https://app.coderabbit.ai/change-stack/invakid404/baml-rest/pull/269)

### CODE REVIEW

File: bamlutils/bamlparser/ast.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// parser tolerates the full upstream BAML surface without erroring).
+type File struct {
+	Items []*Item
+}
+
+// Item is a single top-level declaration. Exactly one of the typed pointers
+// is non-nil for any well-formed Item. Other is used for tolerated upstream
+// constructs the package doesn't model (class, enum, type alias, test, raw
+// top-level assignments, etc.) — its body is captured opaquely so it can be
+// skipped without leaking into the recognised-block walkers.
```
[the remark this answers is not in the document]
Comment (invakid404) [reply bamlutils/bamlparser/ast.go]: @coderabbitai Good catch — fixed in eb09364f2.

The doc comments lagged behind the implementation. After Codex's scoping landed dedicated `TypeBlock`/`TypeAlias`/`TemplateBlock` nodes, the wording on `Item`/`Other` was never updated to reflect that class/enum/test now dispatch to `TypeBlock` (not `Other`), type aliases to `TypeAlias`, and `template_string` to `TemplateBlock`. The "captured opaquely" / "held opaquely" phrasing was also stale — the parser consumes brace bodies and expression right-hand-sides with balanced-skip, never retaining them on the AST.

Rewrote the package doc + Item, TypeBlock, TypeAlias, and Other doc comments to describe the actual model:
- Item lists which typed pointer represents each `.baml` construct, and notes that the tolerated declarations are metadata-only.
- TypeBlock / TypeAlias / Other docs each say explicitly that the brace body or RHS expression is consumed and not retained, and that consumers needing a verbatim copy must read the source bytes directly.
- Other's `Keyword` field doc now correctly says it carries the leading token's value either as an identifier or as the raw token when the item began with punctuation (the prior "or empty when the item began with something other than an identifier" was wrong).

Pure doc fix; no behavioural change. Compiled clean (`go build ./...` exit 0). Tests untouched.

File: cmd/introspect/baml_parser_parity_test.go [hunk 2]
```diff
@@ -0,0 +1,1324 @@
...
+
+	"github.com/invakid404/baml-rest/bamlutils/bamlparser"
+)
+
+// knownParityDivergences enumerates fixtures whose old-parser and
+// new-parser output are not expected to match. Each entry is a fixture
+// name from the parity tables below. The list must remain empty: PR 1
+// asserts the new parser is byte-for-byte equivalent on the documented
+// surface, so any divergence is a regression to escalate.
+var knownParityDivergences = map[string]string{}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/introspect/baml_parser_parity_test.go:37]: @coderabbitai Good catch — fixed in 55639e500.

You're right, the per-case `t.Skipf` was a silent-skip footgun: any future contributor adding an entry to `knownParityDivergences` to make a regression "pass" would do so quietly, completely defeating the harness's purpose. Added a `t.Fatalf` at the top of `TestParity_OldVsNewParser_Fixtures` so a non-empty registry now fails CI loudly before any case runs:

```go
if len(knownParityDivergences) > 0 {
    t.Fatalf("knownParityDivergences must remain empty; parity divergences must be resolved in the parser, not skipped: %#v", knownParityDivergences)
}
```

The per-case `t.Skipf` branch is left in place as defensive documentation (it would short-circuit a single case if reached), but the top-level guard fires first while the invariant holds. The "zero divergences" contract is now load-bearing for PR 2's production switch.

File: cmd/introspect/baml_parser_parity_test.go [hunk 3]
```diff
@@ -0,0 +1,1324 @@
...
+
+function Foo(x: string) -> string {
+    client First
+    prompt #"{{ x }}"#
+    client Second
+}
+`,
+		},
+	}
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/introspect/baml_parser_parity_test.go:1248]: @coderabbitai Real duplication, but deferring this to PR 3 of the #265 sequence rather than addressing in PR 1.

Quantified the overlap: ~51 of 65 parity cases mirror existing `baml_parser_test.go` fixtures (whole-file `parseBamlFile` cases like `ClientAndFunction`, `RetryPolicyConstantAndExponential`, the fallback/RR chain cases, Bedrock endpoint/credentials fixtures); the other ~14 are parity-specific Q1 edge cases hand-built during scoping. The remaining ~14 are either net-new or shape-incompatible.

Shape mismatch makes dedup harder than a pure extract:

- `parityCase` is `{name, src}` and asserts full-`bamlConfig` equality.
- Existing `baml_parser_test.go` tests are not table-driven from a common shape — they allocate per-test config, sometimes preseed stale state, call different entry points (`parseBamlFile`, `parseClientBlock`, `parseStrategyList`, `extractBedrockOptionValue`), and assert targeted fragments rather than whole-config equality. Several also use `[]string` block bodies + a client name rather than complete `.baml` files.

So dedup-now would require either (a) wrapping helper-shape fixtures into full BAML documents and rewriting the helper tests, or (b) inventing a richer fixture API with per-fixture assertion closures. Both are real reshaping work.

PR 3 of the #265 sequence is where this naturally fits: PR 3 deletes the old parser helpers (`stripInlineComment`, `stripStringLiterals`, `cleanBamlValue`, `bamlConfigKeys`, `parseClientBlock`, etc.) AND retargets/removes the helper-specific tests in `baml_parser_test.go`. That's exactly when the test-file shapes get reorganized; folding fixture dedup into that work avoids churning this PR's side-by-side parity harness for infrastructure that PR 3 will rebuild anyway.

I'll include the dedup as an explicit work item when scoping PR 3 (likely shape: `type bamlParserFixture struct { Name, Src string; Assert func(*testing.T, *bamlConfig) }` shared between the parity-survivors and the retargeted introspect tests).
